Proposal: add commit message conventions

oxy opened the pull request:
Add commit message conventions adopted from AngularJS. Objectives:
- improve readability of commit history
- discourage non-descript commit names
- standardise messages to autogenerate changelogs from commit history

Commit:
docs: add commit message conventions

Add commit message conventions adopted from AngularJS. Objectives:
- improve readability of commit history
- discourage non-descript commit names
- standardise messages to autogenerate changelogs from commit history

Martmists-GH commented:
I recommend adding a section about commits that are usually called `minifix` or similar, as these don't require that big of a change (e.g. typo). Could you expand this to include those cases?

sr229 reviewed (dismissed):
Examples should be thoroughly explained. Please also ask @Frosty for some required grammar checking

sr229 commented:
heck I mentioned the wrong person

oxy commented:
CC @SuperFrosty I think the ping was directed towards you. Requested grammar checking.

Commit:
docs(commits): Clarify intent of footer

sr229 reviewed (approved):
LGTM :shipit:

apal0934 commented on CHAPTER3-COMMIT CONVENTIONS.md:
````diff
@@ -0,0 +1,143 @@
+<footer> (optional)
+```
+
+Any line of the commit message should not exceed 100 characters. This allows the message to be easier to read.
````
"Any line of **a** commit message should not exceed 100 characters. This allows **for** the message to be easier to read."

apal0934 commented on CHAPTER3-COMMIT CONVENTIONS.md:
```diff
@@ -0,0 +1,143 @@
+- refactor: A code change that neither fixes a bug nor adds a feature
+- style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
+- test: Adding missing tests or correcting existing tests
+
```
This one's personal preference, but lowercase following a semicolon, and commas at the end of lines except the last which should be a fullstop.

apal0934 commented on CHAPTER3-COMMIT CONVENTIONS.md:
```diff
@@ -0,0 +1,143 @@
+- fix: A bug fix
+- perf: A code change that improves performance
+- refactor: A code change that neither fixes a bug nor adds a feature
+- style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
```
"Meaning" feels weird here. "Functionality", perhaps? Also clarify the difference between this and refactor.

apal0934 commented on CHAPTER3-COMMIT CONVENTIONS.md:
```diff
@@ -0,0 +1,143 @@
+* use imperative, present tense: “change” not “changed” nor “changes”
+* don't capitalize first letter
+* no single dot (.) at the end
+
```
No contractions, (ironically) needs capitalization, and a single dot is called a period.
Also, "Do not capitalize **the** first letter".

apal0934 commented on CHAPTER3-COMMIT CONVENTIONS.md:
```diff
@@ -0,0 +1,143 @@
+## Message body
+* use imperative, present tense: “change” not “changed” nor “changes”
+* include motivation for the change and contrasts with previous behavior
+
```
Again, capitals.
"Include **the** motivation for the change and **the** contrasts with previous behavior"

apal0934 commented on CHAPTER3-COMMIT CONVENTIONS.md:
```diff
@@ -0,0 +1,143 @@
+### Breaking changes
+
+All breaking changes should be mentioned in footer starting with the words `BREAKING CHANGE:`, with a description of the change, justification and migration notes.
+
```
"in **the** footer"

Commit:
docs(commits): type changes, various grammar fixes

- Update description for refactor and style
- Fix capitalization
- Correct various other grammatical errors

sr229 commented:
Can I get a all clear on this one?

Martmists-GH commented:
My question about hotfixes remains unanswered?
